prime-router/src/main/kotlin/transport/SoapTransport.kt

```diff
@@ -0,0 +1,237 @@
+        retryItems: RetryItems?,
+        context: ExecutionContext,
+        actionHistory: ActionHistory
+    ): RetryItems? {
+        // verify that we have a SOAP transport type for our parameters. I think if we ever fell
+        // into this scenario with different parameters there's something seriously wrong in the system,
+        // but it is good to check.
+        val soapTransportType = transportType as? SoapTransportType
+            ?: error("Transport type passed in not of SOAPTransportType")
+        // verify we have a receiver
+        val receiver = header.receiver ?: error("No receiver defined for report ${header.reportFile.reportId}")
+        // get the external file name to send to the client, if we need it
+        val fileName = header.reportFile.externalName
+        context.logger.info(
+            "Preparing to sending ${header.reportFile.reportId} " +
+                "to ${soapTransportType.soapAction} at ${soapTransportType.endpoint}"
+        )
+        // based on who we are sending this report to, we need to get the credentials, and we also need
+        // to create the actual implementation object based on who we're sending to
+        val credential = lookupCredentials(receiver)
+        // calling the SOAP serializer in an attempt to be somewhat abstract here. this is based on the
+        // namespaces provided to the SOAP transport info. It's kind of BS magical string garbage, but the
+        // reality is that each client could have different objects we have to create for them, and these
+        // objects might nest, and/or have different constructors, and it's really hard to be really generic
+        // about this kind of stuff. So this is some semi-tight coupling we will just have to manage.
+        // And honestly, if a client changes their SOAP endpoint, we'd (probably) need to do some coding
+        // on our end, so incurring the maintenance cost here is (probably) okay.
+        val xmlObject = SoapObjectService.getXmlObjectForAction(soapTransportType, header, context, credential)
+            ?: error("Unable to find a SOAP object for the namespaces provided")
+        // wrap the object in the generic envelope. At least this gets to be generic
+        val soapEnvelope = SoapEnvelope(xmlObject, soapTransportType.namespaces ?: emptyMap())
+
+        return try {
+            // run our call to the endpoint in a blocking fashion
+            runBlocking {
+                launch {
+                    val responseBody = connectToSoapService(
+                        soapEnvelope.toXml(),
+                        soapTransportType.endpoint,
+                        soapTransportType.soapAction,
+                        context,
+                        httpClient ?: createDefaultHttpClient()
+                    )
+                    // update the action history
+                    val msg = "Success: SOAP transport of $fileName to $soapTransportType:\n$responseBody"
```

Comment:
will the response body potentially include PII? or is it just a status message?

Reply:
@cwinters-usds it *shouldn't* contain PII. In the case of PA it returns an HL7 ACK message, which has the message ID and any errors generated, but typically those are supposed to be PII free. But there's no guarantee.